vms/proposervm/block/block.go

```diff
@@ -73,32 +80,82 @@ func (b *statelessBlock) Block() []byte {
 	return b.StatelessBlock.Block
 }
 
+func (b *statelessBlock) VRFSig() []byte {
+	return b.StatelessBlock.VRFSig
+}
+
+func (b *statelessBlock) PChainHeight() uint64 {
+	return b.StatelessBlock.PChainHeight
+}
+
+func (b *statelessBlock) Timestamp() time.Time {
+	return b.timestamp
+}
+
+func (b *statelessBlock) Proposer() ids.NodeID {
+	return b.proposer
+}
+
 func (b *statelessBlock) Bytes() []byte {
 	return b.bytes
 }
 
+func (b *statelessBlock) ID() ids.ID {
+	return b.id
+}
+
+func (b *statelessBlock) VerifySignature(pk *bls.PublicKey, parentVRFSig []byte, chainID ids.ID, networkID uint32) bool {
+	if pk == nil {
+		// proposer doesn't have a BLS key.
+		if len(parentVRFSig) == 0 {
+			// parent block had no VRF Signature.
+			// in this case, we verify that the current signature is empty.
+			return len(b.VRFSig()) == 0
+		}
+		// parent block had VRF Signature.
+		expectedSignature := NextHashBlockSignature(parentVRFSig)
+		return bytes.Equal(expectedSignature, b.VRFSig())
+	}
+
+	// proposer does have a BLS key.
+	if len(parentVRFSig) == 0 {
+		// parent block had no VRF Signature.
+		msgHash := calculateBootstrappingBlockSig(chainID, networkID)
+		return bls.Verify(pk, b.vrfSig, msgHash[:])
+	}
+	return bls.Verify(pk, b.vrfSig, parentVRFSig)
+}
+
 func (b *statelessBlock) initialize(bytes []byte) error {
-	b.bytes = bytes
+	var err error
 
-	// The serialized form of the block is the unsignedBytes followed by the
-	// signature, which is prefixed by a uint32. So, we need to strip off the
-	// signature as well as it's length prefix to get the unsigned bytes.
-	lenUnsignedBytes := len(bytes) - wrappers.IntLen - len(b.Signature)
-	unsignedBytes := bytes[:lenUnsignedBytes]
-	b.id = hashing.ComputeHash256Array(unsignedBytes)
+	b.bytes = bytes
+	if b.id, err = initializeID(bytes, b.Signature); err != nil {
+		return err
+	}
 
 	b.timestamp = time.Unix(b.StatelessBlock.Timestamp, 0)
 	if len(b.StatelessBlock.Certificate) == 0 {
 		return nil
 	}
 
-	var err error
 	b.cert, err = staking.ParseCertificate(b.StatelessBlock.Certificate)
 	if err != nil {
 		return fmt.Errorf("%w: %w", errInvalidCertificate, err)
 	}
 
 	b.proposer = ids.NodeIDFromCert(b.cert)
+
+	// The following ensures that we would initialize the vrfSig member only when
+	// the provided signature is 96 bytes long. That supports both statelessUnsignedBlockV0 & statelessUnsignedBlock
+	// variations, as well as optional 32-byte hashes stored in the VRFSig.
+	if len(b.StatelessBlock.VRFSig) == bls.SignatureLen {
```

Comment:
Should `len(b.StatelessBlock.VRFSig)` always be either `0` or `bls.SignatureLen`? If so, should we assert that here?

Reply:
when a proposer doesn't have a BLS key, they would hash the previous signature instead of signing it. In that case, the VRFSig would be 32 byte long.

I think that adding another case here that validates that the signature length is in [0,32,96] makes sense here. I'll add that.

Reply:
done